### PR TITLE
[coin-or-cbc] Add new port

### PR DESCRIPTION
**Pull request by jpfeuffer**
Revive #7850

### CODE REVIEW

**Comment by jpfeuffer** [comment]
@Neumann-A Would you mind having a look here? You recently updated all coin recipes. I was trying to use the Buildtools autoconf files, like you did for all other recipes but it fails.

**Comment by Neumann-A** [comment]
You need to figure out where the missing macros should come from and set ac_local accordingly.

**Comment by jpfeuffer** [comment]
Hmm.. sorry but I don't know what you mean.
Honestly, I don't see a reason why this should fail while the other ports run through. It is essentially the same setup. Now even with the same configure options.

**Comment by jpfeuffer** [comment]
Any idea how to get additional debug information from this incredibly obscure build system? Otherwise I think this out-of-scope for me. I unfortunately cannot invest the time to learn autoconf just for this.

**Comment by dg0yt** [comment]
I started an attempt on building coinutils with the official buildsystem last year: #26524
This is from my notes.
~~~
# We cannot run autoreconf without providing extra coin-or build tools
# and the specific compatible version of autotools, cg.
# https://coin-or-tools.github.io/BuildTools/autotools#using-the-correct-version-of-the-autotools
# But we need to update key scripts for arm64-windows and uwp support.
~~~~

File: ports/coin-or-cbc/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,57 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO coin-or/Cbc
+    REF 133f1171f1667c930e46efd350f4be8ad995b231
+    SHA512 0433715b0d08a3b6862b61371cbbf9956bc98b350c61b9fafe6526feb3d90aad04757d380467cc2b6c8014898ca850fa0d2b325849326e16751de1d68d54b270
+    ## PATCHES fix-c1083-error.patch
```
Comment (jimwang118) [inline ports/coin-or-cbc/portfile.cmake:6]: If the fix-c1083-error.patch is not applied to this port, please delete the patch file. Please remove extra blank lines in the file.

Comment (jimwang118) [inline ports/coin-or-cbc/portfile.cmake:6]:
```suggestion
```

File: ports/coin-or-cbc/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+)
+
+
+if(1)
```
Comment (jimwang118) [inline ports/coin-or-cbc/portfile.cmake:10]: The judgment condition here is always true, then the following branch will not enter. Whether content in the branch below should be deleted.

File: ports/coin-or-cbc/vcpkg.json [hunk 3]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "coin-or-cbc",
+  "version-semver": "2.10.10",
+  "description": "Cbc (Coin-or branch and cut) is an open-source mixed integer linear programming solver written in C++.",
+  "homepage": "https://github.com/coin-or/Cgl",
```
Comment (jimwang118) [inline ports/coin-or-cbc/vcpkg.json:5]:
```suggestion
  "homepage": "https://github.com/coin-or/Cbc",
```

File: versions/c-/coin-or-cbc.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{
+  "versions": [
```
Comment (jimwang118) [inline versions/c-/coin-or-cbc.json:2]: You should use the command: `./vcpkg x-add-version coin-or-cbc` to automatically generate this file instead of writing it manually.

**Review by jimwang118 — CHANGES_REQUESTED** [review]

File: ports/coin-or-cbc/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,57 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO coin-or/Cbc
+    REF 133f1171f1667c930e46efd350f4be8ad995b231
```
Comment (jimwang118) [inline ports/coin-or-cbc/portfile.cmake:4]:
```suggestion
    REF "releases/${VERSION}"
```

**Review by jimwang118 — CHANGES_REQUESTED** [review]

**Comment by jimwang118** [comment]
Note: I will be converting your PR to draft status. The above suggested changes are only recommendations. If you are willing to adopt them, please click "ready for review" after making the modifications. If you do not wish to make any changes, please click "ready for review" directly. That way, I can be aware that you've responded since you can't modify the tags.

**Comment by jpfeuffer** [comment]
Yes will do. waiting for some replies to see what the best way forward is and will remove things once decided.

File: ports/coin-or-cbc/Config.cmake.in [hunk 6]
```diff
@@ -0,0 +1,11 @@
+
+@PACKAGE_INIT@
+
+include(CMakeFindDependencyMacro)
+find_dependency(CoinUtils REQUIRED)
```
Comment (ilya-lavrenov) [inline ports/coin-or-cbc/Config.cmake.in:5]: it should not mention `REQUIRED`

File: ports/coin-or-cbc/portfile.cmake (hunk 5 above)
Comment (jpfeuffer) [reply ports/coin-or-cbc/portfile.cmake]: not possible. see other coin ports. on the release branch, autoconf is not compatible with vcpkg.

**Comment by jpfeuffer** [comment]
I have NO IDEA at all why Windows fails. It's exactly the same setup as in the ports from @Neumann-A now.

File: versions/c-/coin-or-cbc.json (hunk 4 above)
Comment (jpfeuffer) [reply versions/c-/coin-or-cbc.json:2]: It *was* generated with that command but it still complains..

**Comment by jpfeuffer** [comment]
Can someone try this on Windows and maybe help me finish this PR? I would very very much appreciate it.

**Comment by dg0yt** [comment]
CI logs for x64-windows:
~~~
configure:17898: checking D:/downloads/tools/msys2/9a1ec3f33446b195/mingw32/bin/pkg-config.exe is at least version 0.16.0
configure:17902: result: yes: 0.29.2
configure:17938: D:/downloads/tools/msys2/9a1ec3f33446b195/mingw32/bin/pkg-config.exe path is "/D/installed/x64-windows/debug/lib/pkgconfig:D:/installed/x64-windows/debug/lib/pkgconfig/;D:/installed/x64-windows/debug/share/pkgconfig/"
configure:17949: checking for package CoinUtils
configure:18152: result: no
configure:18191: error: Required package CoinUtils is not available.
~~~
The file list says there is a `coinutils.pc`. I know windows file systems aren't case-sensitive, but maybe pkg-config looks at the directory entries. 
(But I see that [coin documentation](https://coin-or-tools.github.io/BuildTools/configure#check-for-packages) uses CamelCase.)

**Comment by jpfeuffer** [comment]
Thanks a lot! Does anyone know why the pkconfig path is a weird mix of Unix and Windows paths? Especially with respect to the path separators ":" and ";"

**Comment by dg0yt** [comment]
Does the port or package add a component to this path itself?
On windows there is context switching between msys and windows envirinments, including implicit conversion of parameters and environment variables which are assumed to be paths. msys is the unix environment in this game.

**Comment by dg0yt** [comment]
> Does the port or package add a component to this path itself?

Yes, https://github.com/coin-or/Cbc/blob/600f7d1095fde2e1c6fc93ed54372880d420ff7b/configure#L17936:
~~~sh
  COIN_PKG_CONFIG_PATH="${expanded_libdir}/pkgconfig:${COIN_PKG_CONFIG_PATH}"
~~~
This doesn't consider pkg-config on windows which uses `;` to separate dirs.

**Comment by jpfeuffer** [comment]
Good find! I guess a patch with an ifelse could work around this?

**Comment by dg0yt** [comment]
I wouldn't mind to see it removed entirely. Don't pick anything from other coin-or locations before proper vcpkg packages.

**Comment by jpfeuffer** [comment]
k, now it is looking for glpk since coinutils is always built with glpk and therefore COINUTILS_HAS_GLPK is active.

@dg0yt Is it possible to configure this package based on the enabled features in the dependency coinutils, after https://github.com/microsoft/vcpkg/pull/31739 was merged?

**Comment by dg0yt** [comment]
.. which triplet?

**Comment by jpfeuffer** [comment]
Ok, Windows x64 builds when deactivating glpk, which I do not plan to support as a feature for now.
Sorry, but I did not completely understand your suggestion @dg0yt and I think it would be a bigger change in this very complex build system. I linked an issue that I raised on the source repo, above.

Lastly I have for the Windows static builds:
`ar-lib: unknown action specified`

**Comment by dg0yt** [comment]
> Ok, Windows x64 builds when deactivating glpk, which I do not plan to support as a feature for now. Sorry, but I did not completely understand your suggestion @dg0yt and I think it would be a bigger change in this very complex build system. I linked an issue that I raised on the source repo, above.

I don't mind if glpk is missing when the port is added.

> Lastly I have for the Windows static builds: `ar-lib: unknown action specified`

Can you try to add `USE_WRAPPERS` to `vcpkg_configure_make`? ATM it is not using the ar-lib  wrapper during configuration tests.

File: ports/coin-or-cbc/portfile.cmake [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    PATCHES "dep.patch" "pkgconf_win.patch" "disable_glpk.patch"
+)
+
+file(COPY "${CURRENT_INSTALLED_DIR}/share/coin-or-buildtools/" DESTINATION "${SOURCE_PATH}")
+
+set(ENV{ACLOCAL} "aclocal -I \"${SOURCE_PATH}/BuildTools\"")
+
+vcpkg_configure_make(
+SOURCE_PATH "${SOURCE_PATH}"
+#AUTOCONFIG
```
Comment (jpfeuffer) [inline ports/coin-or-cbc/portfile.cmake:15]:
```suggestion
USE_WRAPPERS
```

**Review by jimwang118 — DISMISSED** [review]

File: ports/coin-or-cbc/portfile.cmake [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    --with-coinutils
+    --with-clp
+    --with-cgl
+    --with-osi
+    --without-ositests
+    --without-sample
+    --without-netlib
+    --without-miplib3
+    --enable-relocatable
+    --disable-readline
```
Comment (LilyWangLL) [inline ports/coin-or-cbc/portfile.cmake:26]:
```suggestion
    SOURCE_PATH "${SOURCE_PATH}"
    USE_WRAPPERS
    OPTIONS
        --with-coinutils
        --with-clp
        --with-cgl
        --with-osi
        --without-ositests
        --without-sample
        --without-netlib
        --without-miplib3
        --enable-relocatable
        --disable-readline
```

File: ports/coin-or-cbc/portfile.cmake [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+)
+
+vcpkg_install_make()
+vcpkg_copy_pdbs()
+vcpkg_fixup_pkgconfig()
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
+
+file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)
```
Comment (LilyWangLL) [inline ports/coin-or-cbc/portfile.cmake:36]:
```suggestion
vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE")
```

File: ports/coin-or-cbc/portfile.cmake [hunk 10]
```diff
@@ -0,0 +1,36 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO coin-or/Cbc
+    REF 6f83969e50b6f67c60654704c1f71f970c621a3b
+    SHA512 f56b5806b2acffe7259410254640009f9b7a27713972d771b1083a05ca197a65715b007f42f0b8bf6e87b2a889e2889a7466222e0b06f01a4c5297cbaf455c4d
+    PATCHES "dep.patch" "pkgconf_win.patch" "disable_glpk.patch"
```
Comment (LilyWangLL) [inline ports/coin-or-cbc/portfile.cmake:6]:
```suggestion
    PATCHES
        dep.patch
        pkgconf_win.patch
        disable_glpk.patch
```

File: ports/coin-or-cbc/vcpkg.json [hunk 11]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "coin-or-cbc",
+  "version-date": "2023-02-01",
```
Comment (LilyWangLL) [inline ports/coin-or-cbc/vcpkg.json:3]: I saw the upstream has release version [releases/2.10.10](https://github.com/coin-or/Cbc/releases/tag/releases%2F2.10.10), why not use release version?

**Review by LilyWangLL — CHANGES_REQUESTED** [review]

**Comment by LilyWangLL** [comment]
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

File: ports/coin-or-cbc/vcpkg.json (hunk 11 above)
Comment (jpfeuffer) [reply ports/coin-or-cbc/vcpkg.json:3]: As far as I understood, it's too old to work with newer autoconf versions. I used the commit with the autoconf changes to the whole coinor ecosystem. They seem to have been applied on the same day for all coin packages, so it seemed fitting to use that commit here as well.

I think in the original PR for the revamping of the coinor packages this is explained better.

I also raised an issue at their repo and they said there is no release in sight, that we could wait for.

Comment (dg0yt) [reply ports/coin-or-cbc/vcpkg.json:3]: Hm, I'm lost in the the chickens and eggs...

The PR no longer uses the `AUTOCONFIG` parameter, so it is *not* meant to use `autoreconf` (autoconf).
There is a patch to `configure.ac` which doesn't have an effect without `autoreconf`. But it might trigger `autoreconf` from `make` if the patched `configure.ac` is newer than `configure`. I guess this only prevented by the patch to `configure`....

Long story short, isn't it possible to drop the patch to `configure.ac` and build the latest release?

BTW I think there should be `DETERMINE_BUILD_TRIPLET` if `AUTOCONFIG` isn't used. It is needed for cross-builds.

Comment (jpfeuffer) [reply ports/coin-or-cbc/vcpkg.json:3]: Hmm ok makes sense. I did not know how these files and tools work together. I mostly tried to replicate what was done for the other coinor ports.

I can certainly try this. Just a question, since this confused me as well:
Did you consider that these coinor ports including mine, checkout the coinor Buildtools port to get some additional scripts or something? It felt like this is another reason why not the latest release was chosen but I am not sure.
https://github.com/coin-or-tools/BuildTools

Comment (jpfeuffer) [reply ports/coin-or-cbc/vcpkg.json:3]: Ok I added back AUTOCONFIG but I think all the patches are needed. One fixes paths on windows, another allows inclusion of additional macros from the buildtools package and the last one disables glpk.
I don't think that can be solved by going back to an even older codebase. Keep in mind that this package heavily depends on other coin-or packages and all of them are built from this specific point in time when the coin-or team updated their configure scripts.
Maybe @Neumann-A can comment a bit more on the reasons.

**Comment by jpfeuffer** [comment]
Keeping logs:
With AUTOCONFIG + dep.patch for configure.ac:

```
autoreconf: export WARNINGS=
autoreconf: Entering directory '.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal -I "/mnt/vcpkg-ci/buildtrees/coin-or-cbc/src/970c621a3b-821838d561.clean/BuildTools" --force 
aclocal: error: couldn't open directory 'm4': No such file or directory
autoreconf: error: aclocal failed with exit status: 1
```

**Comment by dg0yt** [comment]
To make it clear, I suggest the combination of:  
not AUTOCONFIG
DETERMINE_BUILD_TRIPLET
not patching configure.ac

**Comment by jpfeuffer** [comment]
Can someone retrigger stalled CI?

**Review by LilyWangLL — APPROVED** [review]

**Comment by jpfeuffer** [comment]
Hi! I was wondering if there is something missing. In any case, just bumping so this won't be forgotten.

**Comment by jpfeuffer** [comment]
Great! I want to thank everyone here for their help and that project in general!
